### PR TITLE
Feature/#5 add unit test support

### PR DESCRIPTION
**Pull request by brauliodiez**
Added unit test support to trackerHoc and trackHttp (Jest + React 16)

trackHttp tests not 100% completed, issue with testing catch (will open a specific case for this).

@arp82 once this pull is merged you could add unit test support to the emitter refactor you have proposed (and extend the updated to the code that uses the emitter).

@fjcalzado once we are done with this pull we can open a branch and rename _trackHttp_ to _trackPromise_

We are quite near to have a 1.0 version working :)

### CODE REVIEW

File: src/trackHttp.test.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import {trackPromise, emitter} from './trackHttp';
+import {Emitter} from './tinyemmiter';
+
+
+describe('trackPromise', () => {
+  test('On Initial case, promise fired, promise emitter.emit is called', () => {
```
Comment (arp82) [inline src/trackHttp.test.js]: I understand that the idea of this unit test is to check that .emit() is called once we start tracking a given promise, and it should not be called again until that promise is resolved/rejected. In light of this, I would rename it to something like "when tracking a single, still pending promise, emitter.emit should be called once"

File: src/trackHttp.test.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    // Arrange
+    emitter.emit = jest.fn((a,b) => {
+      return;
+    });
+
+    const myPromise =  Promise.resolve().then(() => {
```
Comment (arp82) [inline src/trackHttp.test.js]: The idea here is that this promise is not going to be resolved by the time the "expect" clause is evaluated. Thus, I would rename this to myPendingPromise or myDeferredPromise for better clarity. 
(On a side note, I really miss the deprecated Promise.defer() method here, it feels rather artificial and convoluted to do it this way, but I cannot think of a better way either).

File: src/trackHttp.test.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    // Act
+    trackPromise(myPromise);
+
+    // Assert
+    expect(emitter.emit).toHaveBeenCalledTimes(1);
+
+    return myPromise;
+  });
+
+  test('Promise tracked, we got resolve, check that emit is called 2 times', () => {
```
Comment (arp82) [inline src/trackHttp.test.js]: I would rename it to: "when tracking a single promise and after it resolves, emit should have been called twice"

File: src/trackHttp.test.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    // Act
+    trackPromise(myPromise);
+
+    // Assert
+    return myPromise.then(() => {
+      expect(emitter.emit).toHaveBeenCalledTimes(2);
+    });
+  });
+
+  test.skip('Promise tracked, we got fail, check that emit is called 2 times', () => {
```
Comment (arp82) [inline src/trackHttp.test.js]: I would rename it to: "when tracking a single promise and after it is rejected, emit should have been called twice"

File: src/trackHttp.test.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    // Assert
+    return myPromise.catch(()=> {
+        expect(emitter.emit).toHaveBeenCalledTimes(2)
+    });
+  });
+
+
+  // Pending promise failed
+  test('Two Promises tracked, we got resolve on both, check that emit is called 4 times', () => {
```
Comment (arp82) [inline src/trackHttp.test.js]: In a similar vein to previous cases, I would rename this one to: "when tracking two promises and after both are resolved successfully, emit should have been called four times"

**Review by arp82 — APPROVED** [review]
Minor comments regarding clarity in the tests names. The tests seem to work fine and their execution matches the expected behaviour of the trackPromise and promiseTrackerHoc methods

**Comment by fjcalzado** [comment]
> @fjcalzado once we are done with this pull we can open a branch and rename trackHttp to trackPromise

@brauliodiez `trackHttp` file was renamed and already committed in #10.  This branch seems to be in conflict for that reason. Do you want me to fix it?

**Comment by fjcalzado** [comment]
@brauliodiez I felt free to fix this conflict by merging changes from #10 . Also, I had to do some minor fixes in the build system to make it work.

**Comment by brauliodiez** [comment]
Merged, there are some edge cases we wil have to discusss, not sure how jest can handle them

**Comment by brauliodiez** [comment]
1.0 release is soon to be reached

**Comment by fjcalzado** [comment]
Sure, let me know and we can add them. I had been thinking in some extra tests today as well. I will try to spend some time tomorrow so we can add the linter and whatever thing you consider important.

**Pull request merged**